src/main/java/org/jenkinsci/plugins/workflow/graph/FlowNode.java

```diff
@@ -244,66 +247,120 @@ public String getUrl() throws IOException {
      *
      * This method provides such an opportunity for subtypes.
      */
-    protected void setActions(List<Action> actions) {
-        this.actions = new CopyOnWriteArrayList<Action>(actions);
+    protected synchronized void setActions(List<Action> actions) {
+            this.actions = new CopyOnWriteArrayList<Action>(actions);
+//            this._wrapper = new ListWrapper(this);
     }
 
-/*
-    We can't use Actionable#actions to store actions because they aren't transient,
-    and we need to store actions elsewhere because this is the only mutable pat of FlowNode.
+    /**
+     * Return the first persistent action on the FlowNode, without consulting TransientActionFactories
+     * Used here because it is much faster to invoke a less-conditional, monomorphic method (50% faster)
+     * @param type Class of action
+     * @param <T>  Action type
+     * @return First persistent action or null if not found
+     */
+    public <T extends Action> T getPersistentAction(@Nonnull Class<T> type) {
```

Comment:
I feel like you don't need this above `private` scope at all. You can just override `getAction(Class<T> type)` e.g.

```
public <T extends Action> T getAction(@Nonnull Class<T> type) {
  if (PersistentAction.class.isAssignableFrom(type)) {
    // short-cut
    if (actions == null) {
      loadActions();
    }
     for (Action a : actions) {
       if (type.isInstance(a)) {
         return type.cast(a);
       }
     }
     // no point checking super as we know its a persistent action
     return null;
  }
  // we have to check the transients now
  return super.getAction(type);
}
```

and that should be short enough to fit within inlining without requiring callers to know which getAction method to call